### PR TITLE
Use more specific locations for JSONSelection when available

### PR DESCRIPTION
**Pull request by dylan-apollo**
Also reorganizes selection code a bit with some new helper structs.

### CODE REVIEW

**Review by benjamn — APPROVED** [review]
Approving based on your screen recording and a quick review of JSONSelection API use. Other folks should also review, probably.

File: apollo-federation/src/sources/connect/validation/selection.rs [hunk 1]
```diff
@@ -109,61 +101,49 @@ pub(super) fn validate_body_selection(
...
-        locations: selection_arg
-            .value
-            .line_column_range(source_map)
-            .into_iter()
-            .collect(),
-    })?;
+    let (_rest, selection) =
+        JSONSelection::parse(selection_str.as_str()).map_err(|err| Message {
+            code: Code::InvalidJsonSelection,
+            message: format!("{coordinate} is not a valid JSONSelection: {err}",),
```
[the remark this answers is not in the document]
Comment (dylan-apollo) [reply apollo-federation/src/sources/connect/validation/selection.rs:135]: Yeah—though probably the safest thing is to impl `std::error::Error`, which requires `Display + Debug` (the functions in the trait itself should be optional).

[the remark this answers is not in the document]
Comment (benjamn) [reply apollo-federation/src/sources/connect/validation/selection.rs:135]: Ok cool, that's what I did in #6089.

**Review by lennyburdette — APPROVED** [review]

File: apollo-federation/src/sources/connect/validation/snapshots/validation_tests@denest_scalars.graphql.snap [hunk 2]
```diff
@@ -8,7 +8,7 @@ input_file: apollo-federation/src/sources/connect/validation/test_data/denest_sc
         code: GroupSelectionRequiredForObject,
         message: "`User.street` is an object, so ``@connect(selection:)` on `Query.me`` must select a group `street{}`.",
```
Comment (benjamn) [inline apollo-federation/src/sources/connect/validation/snapshots/validation_tests@denest_scalars.graphql.snap:9]:
```suggestion
        message: "`User.street` is an object, so `@connect(selection:)` on `Query.me` must select a group `street{}`.",
```
Are the double backticks in this (existing) error message worth fixing?

Comment (lennyburdette) [reply apollo-federation/src/sources/connect/validation/snapshots/validation_tests@denest_scalars.graphql.snap:9]: addressing this in the merge pr https://github.com/apollographql/router/pull/6091/commits/a0338b9065aaa749443f677b0f54bf0a9b040708

**Pull request merged**